### PR TITLE
refactor: use click in dependency_resolver.py

### PR DESCRIPTION
**Pull request by cj81499**
Using click makes it easier to parse arguments. Many args are now named arguments
(options), and the need for using positional args with stub `"None"` values isn't
necessary anymore.

There is already a dependency on click via piptools, so this doesn't introduce a new
dependency.

Relates to #1067

### CODE REVIEW

**Comment by cj81499** [comment]
@daveboutcher This is ready for your review

**Comment by cj81499** [comment]
@hrfuller are you available to take a look at this? If not, is there someone else who might be able to?

**Comment by hrfuller** [comment]
> are you available to take a look at this? If not, is there someone else who might be able to?
Yes, I can take a look. @rickeylev is a more active maintainer than me these days so I'd like him to take a look as well.

File: python/pip_install/tools/dependency_resolver/dependency_resolver.py
Comment (hrfuller) [inline python/pip_install/tools/dependency_resolver/dependency_resolver.py:134]: It looks like we lose sys.argv[0] compared to before. I'm not sure if the piptools cli expects argv to have argv[0] be the name of the original script but maybe we can add sys.argv[0] here in case.

File: python/pip_install/tools/dependency_resolver/dependency_resolver.py
Comment (hrfuller) [inline python/pip_install/tools/dependency_resolver/dependency_resolver.py:164]: Unclear why this was here. probably for debugging?

File: python/pip_install/tools/dependency_resolver/dependency_resolver.py
Comment (cj81499) [reply python/pip_install/tools/dependency_resolver/dependency_resolver.py:134]: Just to be clear, are you suggesting we change this to the following?

```suggestion
    argv = [sys.argv[0]]
```

File: python/pip_install/tools/dependency_resolver/dependency_resolver.py
Comment (cj81499) [reply python/pip_install/tools/dependency_resolver/dependency_resolver.py:164]: That was my assumption too. Didn't seem like something that ought to stick around (adding clutter to stdout!).

Happy to add it back if someone has a good reason for it to stay there.

**Review by cj81499 — COMMENTED** [review]
> @rickeylev is a more active maintainer than me these days so I'd like him to take a look as well.

@hrfuller I take this to mean you'd like me to hold off on clicking the merge button until they review?

**Comment by cj81499** [comment]
@hrfuller @rickeylev bump!

I would really like to get this (and the related #1067) merged.

**Comment by cj81499** [comment]
@hrfuller @rickeylev I'd still like to get this (and #1067) merged.
Please take a look at your earliest convenience.

File: python/pip_install/tools/dependency_resolver/dependency_resolver.py
Comment (rickeylev) [reply python/pip_install/tools/dependency_resolver/dependency_resolver.py:134]: From what I can tell, omitting argv[0] when calling cli() is correct.

Under the hood, piptools is using click, and the implementation is obscured by a ton of decorators. Digging into it...
* `compile.cli` is an instance of the click `Command` class.
* `Command.__call__` is an alias for `Command.main`
* `Command.main` docs say, for the first positionl arg (named `args`): "the arguments that should be used for parsing. If not provided sys.argv[1:] is used"

So it doesn't expect args to contain argv[0]. The docs go on further to say the optional `prog_name` arg is initialized to sys.argv[0] if not specified.

As a side note, there is a `standalone_mode` arg, which controls whether the command tries to exit or not.

**Review by rickeylev — APPROVED** [review]

**Comment by rickeylev** [comment]
Hi @cj81499,

Thanks for your PR. Sorry it took so long to be reviewed. I had let this languish because I thought it was introducing a new dependency on click, but we already have that dependency, so this isn't as bad as I thought.

**Comment by cj81499** [comment]
@rickeylev thank you for getting around to this! ❤️

fwiw, I did say the following in the PR description ;)

> There is already a dependency on click via piptools, so this doesn't introduce a new dependency.

Excited that #1067 will be unblocked once this makes its way through the queue!

**Pull request merged**
